### ISSUE TITLE
Device filters to narrow the search to only one item

### ISSUE THREAD

**Issue by oneingan** [awlsring/terraform-provider-headscale#3]
AFAIK the correct way to use data sources filters is to narrow the search until obtain a unique result, which can be accessed directly. Also timeout/retry parameter when the list of objects is null could be interesting.

Whats your general thought on this? Is this a bug, or maybe is the desirable behaviour and then is a feature.

Thanks for your provider, I'm starting to use and abuse it a lot.

**Comment by awlsring** [comment]
Glad its working for you so far!

When I wrote this, the mindset on list data sources was "I want to get a list of thing", with filters being "I want a list of these certain things". This was inspired seeing how a few other providers handle list with filters. I'm not sure I've seen exactly what you are describing, can you link to an example? I'm not sure with the attributes exposed it'd always be possible to filter down to one result. It could be possible to refactor the single item data source to do this by expanding the logic in the class to look for something based on a combination of different attributes. Can you share use cases you had in mind?

Timeouts for everything would be a good addition, I'd have to look into how to implement them. I might pick this up when I do an update for v0.23.

**Comment by oneingan** [comment]
Sure, I misunderstand the filters section in other providers. I will close the issue.

**Comment by oneingan** [comment]
I finaly have found a provider filtering datasources:

https://registry.terraform.io/providers/cloudflare/cloudflare/latest/docs/data-sources/zone

**Comment by awlsring** [comment]
Yeah, implementing the manner the cloudflare provider does make sense to me.

Looking at this pattern, singular data source typically will accept one of two inputs, an id for the resource or a filter object that helps determining a singular data source. I will need to play with a provider that does this, though I suspect behavior is the data source will error out if none or multiple match the filter.

I'll start to scope out how to approach this. The two data sources that seem applicable for this is device and user.

**Comment by awlsring** [comment]
I partially implemented this for user in release 0.3.0. A user data source now optionally allows setting any of its attributes which will be used in the API call of ListUsers. 

I'll be adding support for v0.26.0 next, but after that I'll see the feasibility of adding this for devices

**Comment by awlsring** [comment]
In release 0.4.0, I've implemented something close to this ask, where a device data source can now be selected based off a few other "unique" identifiers (given name, name)

I think this is about as far as I'd like to add this. If this doesn't work with your usecase, feel free to reopen and provide some more examples of what you are thinking.
